mud-contracts/smart-object-framework-v2/src/WorldWithContext.sol

```diff
@@ -0,0 +1,589 @@
+    bytes32 trackedSystemId;
+    bytes32 trackedFunctionSelector;
+    bytes32 trackedMsgSender;
+    uint256 trackedMsgValue;
+
+    assembly {
+      let callCount := tload(0)
+      trackedSystemId := tload(add(mul(sub(callCount, 1), 4), 1))
+      trackedFunctionSelector := tload(add(mul(sub(callCount, 1), 4), 2))
+      trackedMsgSender := tload(add(mul(sub(callCount, 1), 4), 3))
+      trackedMsgValue := tload(add(mul(sub(callCount, 1), 4), 4))
+    }
+
+    ResourceId systemId = ResourceId.wrap(trackedSystemId);
+    bytes4 functionSelector = bytes4(trackedFunctionSelector);
+    address msgSender = address(uint160(uint256(trackedMsgSender)));
+
+    return (systemId, functionSelector, msgSender, uint256(trackedMsgValue));
+  }
+
+  /**
+   * @notice Retrieves execution context details for a specific world call
+   * @dev Reads system ID, function selector, msg.sender and msg.value from transient storage at specified call count
+   * @param callCount The specific call count to retrieve execution context values for
+   * @return ResourceId The system ID that was called
+   * @return bytes4 The function selector that was called
+   * @return address The msg.sender of the call
+   * @return uint256 The msg.value of the call
+   */
+  function getWorldCallContext(uint256 callCount) public view returns (ResourceId, bytes4, address, uint256) {
+    bytes32 trackedSystemId;
+    bytes32 trackedFunctionSelector;
+    bytes32 trackedMsgSender;
+    uint256 trackedMsgValue;
+
+    assembly {
+      trackedSystemId := tload(add(mul(sub(callCount, 1), 4), 1))
+      trackedFunctionSelector := tload(add(mul(sub(callCount, 1), 4), 2))
+      trackedMsgSender := tload(add(mul(sub(callCount, 1), 4), 3))
+      trackedMsgValue := tload(add(mul(sub(callCount, 1), 4), 4))
+    }
+
+    ResourceId systemId = ResourceId.wrap(trackedSystemId);
+    bytes4 functionSelector = bytes4(trackedFunctionSelector);
+    address msgSender = address(uint160(uint256(trackedMsgSender)));
+
+    return (systemId, functionSelector, msgSender, uint256(trackedMsgValue));
+  }
+
+  /**
+   * @notice Gets the current depth of the world execution call stack
+   * @dev Reads the call counter from slot 0 in transient storage
+   * @return uint256 The total number of world calls invoked/tracked in the current transaction
+   */
+  function getWorldCallCount() public view returns (uint256) {
+    uint256 callCount;
+    assembly {
+      callCount := tload(0)
+    }
+    return callCount;
+  }
+
+  function _trackWorldCallContextBefore(address msgSender, ResourceId systemId, bytes4 functionSelector) internal {
+    bytes32 trackedSystemId = ResourceId.unwrap(systemId);
+    bytes32 trackedMsgSender = bytes32(uint256(uint160(msgSender)));
+    uint256 trackedMsgValue = msg.value;
+    assembly {
+      let callCount := tload(0)
+      let sysIdSlot := add(mul(callCount, 4), 1)
+      let functIdSlot := add(mul(callCount, 4), 2)
```

Comment:
is it required to calculate the slot index  each time. Since we know its 4 slots can't we increment the slot index ?

Reply:
You could be right, it may be possible to accomplish this through incrementing. However, for a first cut implementation I implemented the current way to be absolutely certain that we are using the current call count accurately in our calculations. So essentially, erring on the side of guaranteed expected behavior. 

But we can revisit this later for sure for optimizing!